File: ServerCore/src/main/java/org/elkoserver/foundation/server/BaseTcpConnectionSetup.java
package org.elkoserver.foundation.server;

import org.elkoserver.foundation.net.MessageHandlerFactory;
import org.elkoserver.foundation.net.NetAddr;
import org.elkoserver.foundation.net.NetworkManager;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.foundation.timer.Timer;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

import java.io.IOException;

abstract class BaseTcpConnectionSetup extends BaseConnectionSetup {
    private String hostIncludingPortNumber;
    final NetworkManager myNetworkManager;
    final MessageHandlerFactory actorFactory;

    BaseTcpConnectionSetup(String label, String host, AuthDesc auth, boolean secure, ElkoProperties props, String propRoot, NetworkManager networkManager, MessageHandlerFactory actorFactory, Trace trServer, Trace tr, TraceFactory traceFactory) {
        super(label, host, auth, secure, props, propRoot, trServer, tr, traceFactory);

        hostIncludingPortNumber = host;
        myNetworkManager = networkManager;
        this.actorFactory = actorFactory;
    }

    @Override
    public String getServerAddress() {
        return hostIncludingPortNumber;
    }

    @Override
    NetAddr tryToStartListener() throws IOException {
        NetAddr result = createListenAddress();
        if (host.indexOf(':') < 0) {
            hostIncludingPortNumber = host + ":" + result.getPort();
        }
        return result;
    }

    abstract NetAddr createListenAddress() throws IOException;

    @Override
    String getListenAddressDescription() {
        return hostIncludingPortNumber;
    }

    @Override
    String getValueToCompareWithBind() {
        return hostIncludingPortNumber;
    }
}

File: ServerCore/src/main/kotlin/org/elkoserver/foundation/server/BaseTcpConnectionSetup.kt
package org.elkoserver.foundation.server

import org.elkoserver.foundation.net.MessageHandlerFactory
import org.elkoserver.foundation.net.NetAddr
import org.elkoserver.foundation.net.NetworkManager
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.metadata.AuthDesc
import org.elkoserver.util.trace.TraceFactory
import org.elkoserver.util.trace.slf4j.Gorgel
import java.io.IOException

abstract class BaseTcpConnectionSetup(
        label: String?,
        override var serverAddress: String,
        auth: AuthDesc,
        secure: Boolean,
        props: ElkoProperties,
        propRoot: String,
        val myNetworkManager: NetworkManager,
        val actorFactory: MessageHandlerFactory,
        gorgel: Gorgel,
        listenerGorgel: Gorgel,
        traceFactory: TraceFactory)
    : BaseConnectionSetup(label, serverAddress, auth, secure, props, propRoot, gorgel, listenerGorgel, traceFactory) {

    @Throws(IOException::class)
    override fun tryToStartListener(): NetAddr {
        val result = createListenAddress()
        if (serverAddress.indexOf(':') < 0) {
            serverAddress = "$serverAddress:${result.port}"
        }
        return result
    }

    @Throws(IOException::class)
    abstract fun createListenAddress(): NetAddr

    override val listenAddressDescription: String
        get() = serverAddress

    override val valueToCompareWithBind: String
        get() = serverAddress
}
